File: sdk/typescript/test_sdk.mjs
/**
 * OculOS TypeScript SDK — integration test (runs as plain JS with Node 22+).
 * We import the .ts source directly won't work without tsx, so we inline the client logic.
 */

const BASE = "http://127.0.0.1:7878";

class OculOS {
  constructor(baseUrl = BASE) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
  }

  async listWindows() { return this._get("/windows"); }
  async getTree(pid) { return this._get(`/windows/${pid}/tree`); }
  async getTreeHwnd(hwnd) { return this._get(`/hwnd/${hwnd}/tree`); }

  async findElements(pid, opts = {}) {
    const p = new URLSearchParams();
    if (opts.query) p.set("q", opts.query);
    if (opts.type) p.set("type", opts.type);
    if (opts.interactive !== undefined) p.set("interactive", String(opts.interactive));
    const qs = p.toString();
    return this._get(`/windows/${pid}/find${qs ? `?${qs}` : ""}`);
  }

  async findElementsHwnd(hwnd, opts = {}) {
    const p = new URLSearchParams();
    if (opts.query) p.set("q", opts.query);
    if (opts.type) p.set("type", opts.type);
    if (opts.interactive !== undefined) p.set("interactive", String(opts.interactive));
    const qs = p.toString();
    return this._get(`/hwnd/${hwnd}/find${qs ? `?${qs}` : ""}`);
  }

  async focusWindow(pid) { await this._post(`/windows/${pid}/focus`); }
  async closeWindow(pid) { await this._post(`/windows/${pid}/close`); }
  async click(id) { await this._post(`/interact/${id}/click`); }
  async setText(id, text) { await this._post(`/interact/${id}/set-text`, { text }); }
  async sendKeys(id, keys) { await this._post(`/interact/${id}/send-keys`, { keys }); }
  async focus(id) { await this._post(`/interact/${id}/focus`); }
  async toggle(id) { await this._post(`/interact/${id}/toggle`); }
  async expand(id) { await this._post(`/interact/${id}/expand`); }
  async collapse(id) { await this._post(`/interact/${id}/collapse`); }
  async select(id) { await this._post(`/interact/${id}/select`); }
  async setRange(id, value) { await this._post(`/interact/${id}/set-range`, { value }); }
  async scroll(id, direction) { await this._post(`/interact/${id}/scroll`, { direction }); }
  async scrollIntoView(id) { await this._post(`/interact/${id}/scroll-into-view`); }
  async highlight(id, durationMs = 2000) { await this._post(`/interact/${id}/highlight`, { duration_ms: durationMs }); }
  async health() { return this._get("/health"); }

  async _get(path) {
    const res = await fetch(`${this.baseUrl}${path}`);
    const body = await res.json();
    if (!body.success) throw new Error(body.error || `HTTP ${res.status}`);
    return body.data;
  }

  async _post(path, json) {
    const res = await fetch(`${this.baseUrl}${path}`, {
      method: "POST",
      headers: json ? { "Content-Type": "application/json" } : {},
      body: json ? JSON.stringify(json) : undefined,
    });
    const body = await res.json();
    if (!body.success) throw new Error(body.error || `HTTP ${res.status}`);
    return body.data;
  }
}

// ── Tests ──

let passed = 0;
let failed = 0;

function ok(name, msg) { console.log(`  ✓ ${name} — ${msg}`); passed++; }
function fail(name, msg) { console.log(`  ✗ ${name} — ${msg}`); failed++; }

async function run() {
  console.log("OculOS TypeScript SDK Test\n");
  const client = new OculOS();

  // 1. health
  try {
    const h = await client.health();
    if (h.status !== "running") throw new Error(`status=${h.status}`);
    ok("health()", `status=${h.status}, version=${h.version}`);
  } catch (e) { fail("health()", e.message); }

  // 2. listWindows
  let pid, hwnd;
  try {
    const wins = await client.listWindows();
    if (!Array.isArray(wins) || wins.length === 0) throw new Error("empty");
    pid = wins[0].pid;
    hwnd = wins[0].hwnd;
    ok("listWindows()", `${wins.length} windows`);
  } catch (e) { fail("listWindows()", e.message); }

  // 3. getTree
  try {
    const tree = await client.getTree(pid);
    if (!tree.oculos_id) throw new Error("no oculos_id");
    ok("getTree()", `root=${tree.type}, children=${tree.children.length}`);
  } catch (e) { fail("getTree()", e.message); }

  // 4. getTreeHwnd
  try {
    const tree2 = await client.getTreeHwnd(hwnd);
    if (!tree2.oculos_id) throw new Error("no oculos_id");
    ok("getTreeHwnd()", `root=${tree2.type}`);
  } catch (e) { fail("getTreeHwnd()", e.message); }

  // 5. findElements
  try {
    const elems = await client.findElements(pid);
    if (!Array.isArray(elems)) throw new Error("not array");
    ok("findElements()", `${elems.length} elements`);
  } catch (e) { fail("findElements()", e.message); }

  // 6. findElements interactive
  try {
    const elems = await client.findElements(pid, { interactive: true });
    ok("findElements(interactive)", `${elems.length} interactive`);
  } catch (e) { fail("findElements(interactive)", e.message); }

  // 7. findElementsHwnd
  try {
    const elems = await client.findElementsHwnd(hwnd, { interactive: true });
    ok("findElementsHwnd()", `${elems.length} interactive`);
  } catch (e) { fail("findElementsHwnd()", e.message); }

  // 8. focusWindow
  try {
    await client.focusWindow(pid);
    ok("focusWindow()", "OK");
  } catch (e) { fail("focusWindow()", e.message); }

  // 9. click
  try {
    const btns = await client.findElements(pid, { type: "Button", interactive: true });
    if (btns.length > 0) {
      await client.click(btns[0].oculos_id);
      ok("click()", `${btns[0].oculos_id.slice(0, 8)}... OK`);
    } else {
      ok("click()", "no buttons, skipped");
    }
  } catch (e) { fail("click()", e.message); }

  // 10. focus element
  try {
    const elems = await client.findElements(pid, { interactive: true });
    if (elems.length > 0) {
      await client.focus(elems[0].oculos_id);
      ok("focus()", `${elems[0].oculos_id.slice(0, 8)}... OK`);
    } else {
      ok("focus()", "skipped");
    }
  } catch (e) { fail("focus()", e.message); }

  // 11. highlight
  try {
    const elems = await client.findElements(pid, { interactive: true });
    if (elems.length > 0) {
      await client.highlight(elems[0].oculos_id, 500);
      ok("highlight()", `${elems[0].oculos_id.slice(0, 8)}... OK`);
    } else {
      ok("highlight()", "skipped");
    }
  } catch (e) { fail("highlight()", e.message); }

  // 12. error handling
  try {
    await client.click("nonexistent-id-12345");
    fail("error handling", "should have thrown");
  } catch (e) {
    if (e.message.includes("not found")) {
      ok("error handling", `Error thrown: '${e.message.slice(0, 50)}...'`);
    } else {
      fail("error handling", `unexpected: ${e.message}`);
    }
  }

  // 13. bad URL
  try {
    const bad = new OculOS("http://127.0.0.1:9999");
    await bad.health();
    fail("bad URL", "should have thrown");
  } catch (e) {
    ok("bad URL", "connection error raised correctly");
  }

  // Summary
  const total = passed + failed;
  console.log(`\n${"=".repeat(40)}`);
  console.log(`  TypeScript SDK: ${passed}/${total} passed`);
  if (failed) console.log(`  ✗ ${failed} FAILED`);
  else console.log(`  ✓ ALL PASSED`);
  console.log("=".repeat(40));
  process.exit(failed ? 1 : 0);
}

run();

File: sdk/python/test_sdk.py
"""OculOS Python SDK — integration test."""

import sys
import os

# Add SDK to path
sys.path.insert(0, os.path.dirname(__file__))

from oculos import OculOS
from oculos.client import OculOSError

def test():
    client = OculOS()
    passed = 0
    failed = 0

    # ── 1. health() ──
    try:
        h = client.health()
        assert h["status"] == "running", f"Expected 'running', got {h['status']}"
        assert "version" in h
        assert "uptime_secs" in h
        print(f"  ✓ health() — status={h['status']}, version={h['version']}")
        passed += 1
    except Exception as e:
        print(f"  ✗ health() — {e}")
        failed += 1

    # ── 2. list_windows() ──
    try:
        windows = client.list_windows()
        assert isinstance(windows, list), "Expected list"
        assert len(windows) > 0, "No windows found"
        w = windows[0]
        assert "pid" in w
        assert "hwnd" in w
        assert "title" in w
        assert "exe_name" in w
        print(f"  ✓ list_windows() — {len(windows)} windows found")
        passed += 1
    except Exception as e:
        print(f"  ✗ list_windows() — {e}")
        failed += 1

    # Pick a window for further tests
    pid = windows[0]["pid"] if windows else None
    hwnd = windows[0]["hwnd"] if windows else None

    # ── 3. get_tree(pid) ──
    try:
        tree = client.get_tree(pid)
        assert tree is not None
        assert "oculos_id" in tree
        assert "type" in tree
        assert "children" in tree
        print(f"  ✓ get_tree({pid}) — root type={tree['type']}, children={len(tree['children'])}")
        passed += 1
    except Exception as e:
        print(f"  ✗ get_tree({pid}) — {e}")
        failed += 1

    # ── 4. get_tree_hwnd(hwnd) ──
    try:
        tree2 = client.get_tree_hwnd(hwnd)
        assert tree2 is not None
        assert "oculos_id" in tree2
        print(f"  ✓ get_tree_hwnd({hwnd}) — root type={tree2['type']}")
        passed += 1
    except Exception as e:
        print(f"  ✗ get_tree_hwnd({hwnd}) — {e}")
        failed += 1

    # ── 5. find_elements(pid) — no filter ──
    try:
        elems = client.find_elements(pid)
        assert isinstance(elems, list)
        print(f"  ✓ find_elements({pid}) — {len(elems)} elements")
        passed += 1
    except Exception as e:
        print(f"  ✗ find_elements({pid}) — {e}")
        failed += 1

    # ── 6. find_elements with query ──
    try:
        elems2 = client.find_elements(pid, interactive=True)
        assert isinstance(elems2, list)
        print(f"  ✓ find_elements(interactive=True) — {len(elems2)} interactive elements")
        passed += 1
    except Exception as e:
        print(f"  ✗ find_elements(interactive=True) — {e}")
        failed += 1

    # ── 7. find_elements_hwnd ──
    try:
        elems3 = client.find_elements_hwnd(hwnd, interactive=True)
        assert isinstance(elems3, list)
        print(f"  ✓ find_elements_hwnd({hwnd}) — {len(elems3)} interactive elements")
        passed += 1
    except Exception as e:
        print(f"  ✗ find_elements_hwnd({hwnd}) — {e}")
        failed += 1

    # ── 8. focus_window ──
    try:
        client.focus_window(pid)
        print(f"  ✓ focus_window({pid}) — OK")
        passed += 1
    except Exception as e:
        print(f"  ✗ focus_window({pid}) — {e}")
        failed += 1

    # ── 9. click — find a button first ──
    element_id = None
    try:
        buttons = client.find_elements(pid, element_type="Button", interactive=True)
        if buttons:
            element_id = buttons[0]["oculos_id"]
            client.click(element_id)
            print(f"  ✓ click({element_id[:8]}...) — OK")
            passed += 1
        else:
            print(f"  ⊘ click — no buttons found, skipped")
            passed += 1
    except Exception as e:
        print(f"  ✗ click — {e}")
        failed += 1

    # ── 10. focus (element) ──
    try:
        if element_id:
            # Re-fetch since tree might have changed
            buttons = client.find_elements(pid, element_type="Button", interactive=True)
            if buttons:
                eid = buttons[0]["oculos_id"]
                client.focus(eid)
                print(f"  ✓ focus({eid[:8]}...) — OK")
                passed += 1
            else:
                print(f"  ⊘ focus — no buttons, skipped")
                passed += 1
        else:
            print(f"  ⊘ focus — no element, skipped")
            passed += 1
    except Exception as e:
        print(f"  ✗ focus — {e}")
        failed += 1

    # ── 11. highlight ──
    try:
        elems_h = client.find_elements(pid, interactive=True)
        if elems_h:
            hid = elems_h[0]["oculos_id"]
            client.highlight(hid, duration_ms=500)
            print(f"  ✓ highlight({hid[:8]}...) — OK")
            passed += 1
        else:
            print(f"  ⊘ highlight — no elements, skipped")
            passed += 1
    except Exception as e:
        print(f"  ✗ highlight — {e}")
        failed += 1

    # ── 12. OculOSError handling ──
    try:
        client.click("nonexistent-id-12345")
        print(f"  ✗ error handling — should have raised OculOSError")
        failed += 1
    except OculOSError as e:
        print(f"  ✓ error handling — OculOSError raised: '{str(e)[:50]}...'")
        passed += 1
    except Exception as e:
        print(f"  ✗ error handling — unexpected: {type(e).__name__}: {e}")
        failed += 1

    # ── 13. Custom base URL ──
    try:
        bad_client = OculOS("http://127.0.0.1:9999")
        bad_client.health()
        print(f"  ✗ bad URL — should have raised")
        failed += 1
    except Exception:
        print(f"  ✓ bad URL — connection error raised correctly")
        passed += 1

    # ── Summary ──
    total = passed + failed
    print(f"\n{'='*40}")
    print(f"  Python SDK: {passed}/{total} passed")
    if failed:
        print(f"  ✗ {failed} FAILED")
    else:
        print(f"  ✓ ALL PASSED")
    print(f"{'='*40}")
    return failed == 0


if __name__ == "__main__":
    print("OculOS Python SDK Test\n")
    ok = test()
    sys.exit(0 if ok else 1)
